google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/model.scala

```diff
@@ -77,66 +77,168 @@ object PubSubConfig {
+                                   val attributes: Option[immutable.Map[String, String]],
+                                   val messageId: String,
+                                   val publishTime: Instant) {
 
   def withAttributes(attributes: java.util.Map[String, String]): PubSubMessage =
-    copy(attributes = Some(attributes.asScala.toMap))
+    new PubSubMessage(data, Some(attributes.asScala.toMap), messageId, publishTime)
 
-  def withPublishTime(publishTime: Instant): PubSubMessage =
-    copy(publishTime = Some(publishTime))
+  def withData(data: String): PubSubMessage =
+    new PubSubMessage(Some(data), attributes, messageId, publishTime)
 
+  override def equals(other: Any): Boolean = other match {
+    case that: PubSubMessage =>
+      data == that.data && attributes == that.attributes && messageId == that.messageId && publishTime == that.publishTime
+    case _ => false
+  }
+
+  override def hashCode: Int = java.util.Objects.hash(data, attributes, messageId, publishTime)
+
+  override def toString: String =
+    "PubSubMessage(data=" + data + ",attributes=" + attributes + ",messageId=" + messageId + ",publishTime=" + publishTime + ")"
 }
 
 object PubSubMessage {
 
-  def apply(data: String): PubSubMessage = PubSubMessage(data, "")
+  def apply(data: Option[String] = None,
+            attributes: Option[immutable.Map[String, String]] = None,
+            messageId: String,
+            publishTime: Instant) = new PubSubMessage(data, attributes, messageId, publishTime)
 
-  def apply(data: String, attributes: immutable.Map[String, String]): PubSubMessage =
-    PubSubMessage(data, "", Some(attributes), None)
-
-  /**
-   * Java API: create [[PubSubMessage]]
-   */
-  def create(data: String) =
-    PubSubMessage(data)
+  def create(data: Option[String],
+             attributes: Option[immutable.Map[String, String]],
+             messageId: String,
+             publishTime: Instant) = new PubSubMessage(data, attributes, messageId, publishTime)
 
-  @deprecated("Setting messageId when creating message for publishing is futile.", "1.1.0")
-  def create(data: String, messageId: String) =
-    PubSubMessage(data, messageId)
 }
 
-final case class PublishRequest(messages: immutable.Seq[PubSubMessage])
+final class PublishRequest private (val messages: immutable.Seq[PublishMessage]) {
+
+  override def equals(other: Any): Boolean = other match {
+    case that: PublishRequest => messages == that.messages
+    case _ => false
+  }
+
+  override def hashCode: Int = messages.hashCode
+
+  override def toString: String = "PublishRequest(" + messages.mkString("[", ",", "]") + ")"
+}
 
 object PublishRequest {
-  def of(messages: java.util.List[PubSubMessage]): PublishRequest =
-    PublishRequest(messages.asScala.toList)
+
+  def apply(messages: immutable.Seq[PublishMessage]): PublishRequest = new PublishRequest(messages)
+
+  def of(messages: java.util.List[PublishMessage]): PublishRequest =
```

Comment:
s/of/create. Now it is a good time to break API.

Reply:
This still needs to be renamed.